tests/unit/test_tracing.c

```diff
@@ -1377,37 +1378,165 @@ SENTRY_TEST(set_trace_id_with_txn)
         = "2674eb52d5874b13b560236d6c79ce8a-a0f9fdf04f1a63df-1";
     const char *txn_trace_id = "2674eb52d5874b13b560236d6c79ce8a";
     const char *txn_parent_span_id = "a0f9fdf04f1a63df";
+    sentry_transaction_context_t *tx_ctx_scoped
+        = sentry_transaction_context_new("wow!", NULL);
+    sentry_transaction_context_update_from_header(
+        tx_ctx_scoped, "sentry-trace", trace_header);
+    sentry_transaction_t *tx_scoped
+        = sentry_transaction_start(tx_ctx_scoped, sentry_value_new_null());
+
+    // when no set_trace was called yet, the scoped transaction should apply
+    //  its trace/parent span ID (as set by update_from_header)
+    sentry_set_transaction_object(tx_scoped);
+    apply_scope_and_check_trace_context(
+        options, txn_trace_id, txn_parent_span_id);
+    sentry_transaction_finish(tx_scoped);
+
     sentry_transaction_context_t *tx_ctx
         = sentry_transaction_context_new("wow!", NULL);
     sentry_transaction_context_update_from_header(
         tx_ctx, "sentry-trace", trace_header);
     sentry_transaction_t *tx
         = sentry_transaction_start(tx_ctx, sentry_value_new_null());
+    sentry_span_t *span_child
+        = sentry_transaction_start_child(tx, "op", "desc");
+    sentry_span_t *span_grandchild
+        = sentry_span_start_child(span_child, "op_g", "desc_g");
 
     // set the direct trace first
     const char *direct_trace_id = "aaaabbbbccccddddeeeeffff00001111";
     const char *direct_parent_span_id = "f0f0f0f0f0f0f0f0";
     sentry_set_trace(direct_trace_id, direct_parent_span_id);
 
     // events should get that trace applied
+    // TODO reason why they should get that trace applied
+    //  (tx started before set_trace, but trace was set so events get that data)
```

Comment:
(idea) Trace was set after Spans started, so these should not have the `trace_id` (unless they now get scoped).

Reply:
This is where it gets complicated and why I believe applying the propagation context when scoping can be over-eager: the application can make sense if it's a transaction, but why would a span be connected to another trace just because the user scoped it?

I know that I proposed this "trace-application at scoping time", but the cleanest approach still is only to do it during creation. It is also a problem that, while we can lock the scope, we can't lock the span (or txn) that a user retains a reference to (so it is also racy).